File: node/src/collaboration/harness.ts
import type { AttemptBody, AttemptBodyContext, AttemptBodyEvent } from "../harness/harness.js"
import type { AttemptJudge, JudgeContext, JudgeResult } from "../harness/judge.js"
import { parseVerdict } from "../runtime/eval.js"
import type { ContractCheckResult, VerificationContract } from "./contract.js"
import { formatContractForSystemPrompt } from "./contract.js"
import type { HandoffArtifact } from "./handoff.js"
import type { AgentPool } from "./pool.js"

export interface ContractOutcome {
  success: boolean
  artifact: string
  checkResults: ContractCheckResult[]
  attemptsUsed: number
  totalTokensConsumed: number
  handoff: HandoffArtifact
}

/** The creator-verifier body owns execution only; verification is an AttemptJudge. */
export class CreatorVerifierBody implements AttemptBody {
  constructor(
    private readonly pool: AgentPool,
    private readonly contract: VerificationContract,
  ) {}

  async *run(context: AttemptBodyContext): AsyncIterable<AttemptBodyEvent> {
    const contractBlock = formatContractForSystemPrompt(this.contract)
    const result = await this.pool.execute("executor", {
      sessionId: context.sessionId,
      goal: `${contractBlock}\n\n---\n\n${context.goal}`,
      ...(context.contextInput ? { contextInput: context.contextInput } : {}),
      verificationContractId: this.contract.id,
    })
    const artifact = result.result.finalMessage?.content ?? ""
    if (artifact) yield { type: "token", text: artifact }
    yield {
      type: "body_done",
      runStatus: String(result.result.termination),
      result: artifact,
      turns: result.result.turnsUsed,
      totalTokens: result.result.totalTokensUsed,
      ...(result.submittedNodes?.length ? { submittedNodes: result.submittedNodes } : {}),
    }
  }
}

/** Structured verifier output only. Free-text PASS/FAIL inference is intentionally unsupported. */
export class StructuredContractJudge implements AttemptJudge {
  constructor(
    private readonly pool: AgentPool,
    private readonly contract: VerificationContract,
  ) {}

  async judge(context: JudgeContext): Promise<JudgeResult> {
    const auditText = await this.pool.verify({ contract: this.contract, artifact: context.result })
    const wire = JSON.parse(auditText) as Record<string, unknown>
    if (typeof wire !== "object" || wire === null || Array.isArray(wire)) {
      throw new Error("structured verifier output must be a JSON object")
    }
    const parsed = parseVerdict(auditText)
    const details = this.contract.acceptance.map(criterion => {
      const detail = parsed.details.find(candidate =>
        candidate.criterion === criterion.id || candidate.criterion === criterion.text)
      return detail
        ? { ...detail, criterion: criterion.id }
        : {
            criterion: criterion.id,
            passed: false,
            score: 0,
            feedback: "criterion missing from structured verifier output",
          }
    })
    const requiredPassed = this.contract.acceptance.every((criterion, index) =>
      !criterion.required || details[index]!.passed)
    return {
      verdict: {
        passed: parsed.passed && requiredPassed,
        overallScore: parsed.overallScore,
        feedback: parsed.feedback,
        details,
      },
    }
  }
}

File: python/deepstrike/collaboration/harness.py
"""Creator/verifier policies consumed by the shared AttemptLoop engine."""

from __future__ import annotations

import json
from dataclasses import dataclass
from typing import TYPE_CHECKING

from deepstrike._kernel import parse_verdict
from deepstrike.collaboration.contract import ContractCheckResult, VerificationContract
from deepstrike.collaboration.contract import format_contract_for_system_prompt
from deepstrike.collaboration.handoff import HandoffArtifact
from deepstrike.harness.harness import (
    AttemptBodyContext,
    AttemptBodyEvent,
    AttemptBodyTerminal,
    AttemptProgressEvent,
    CriterionResult,
    Verdict,
)
from deepstrike.harness.judge import JudgeContext, JudgeResult

if TYPE_CHECKING:
    from deepstrike.collaboration.pool import AgentPool


@dataclass
class ContractOutcome:
    success: bool
    artifact: str
    check_results: list[ContractCheckResult]
    attempts_used: int
    total_tokens_consumed: int
    handoff: HandoffArtifact


class CreatorVerifierBody:
    """Execution-only policy; verification belongs to StructuredContractJudge."""

    def __init__(self, pool: "AgentPool", contract: VerificationContract) -> None:
        self._pool = pool
        self._contract = contract

    async def run(self, context: AttemptBodyContext):
        contract_block = format_contract_for_system_prompt(self._contract)
        result = await self._pool.execute(
            "executor",
            session_id=context.session_id,
            goal=f"{contract_block}\n\n---\n\n{context.goal}",
            context_input=context.context_input,
            verification_contract_id=self._contract.id,
        )
        final = result.result.final_message
        artifact = str(getattr(final, "content", "")) if final else ""
        if artifact:
            yield AttemptProgressEvent("token", {"text": artifact})
        yield AttemptBodyTerminal(
            run_status=result.result.termination,
            result=artifact,
            turns=result.result.turns_used,
            total_tokens=result.result.total_tokens_used,
            submitted_nodes=list(result.submitted_nodes),
        )


class StructuredContractJudge:
    """Accept only the shared structured verdict schema; free-text PASS/FAIL is invalid."""

    def __init__(self, pool: "AgentPool", contract: VerificationContract) -> None:
        self._pool = pool
        self._contract = contract

    async def judge(self, context: JudgeContext) -> JudgeResult:
        from deepstrike.collaboration.pool import IsolatedVerifierContext

        audit_text = await self._pool.run_verifier(
            IsolatedVerifierContext(contract=self._contract, artifact=context.result)
        )
        wire = json.loads(audit_text)
        if not isinstance(wire, dict):
            raise ValueError("structured verifier output must be a JSON object")
        parsed = parse_verdict(audit_text)

        parsed_details = list(parsed.details or [])
        details: list[CriterionResult] = []
        for criterion in self._contract.acceptance:
            match = next(
                (
                    detail
                    for detail in parsed_details
                    if detail.criterion in {criterion.id, criterion.text}
                ),
                None,
            )
            details.append(
                CriterionResult(
                    criterion=criterion.id,
                    passed=bool(match.passed) if match else False,
                    score=float(match.score) if match else 0.0,
                    feedback=(
                        str(match.feedback)
                        if match
                        else "criterion missing from structured verifier output"
                    ),
                )
            )

        required_passed = all(
            not criterion.required or details[index].passed
            for index, criterion in enumerate(self._contract.acceptance)
        )
        return JudgeResult(
            verdict=Verdict(
                passed=bool(parsed.passed) and required_passed,
                overall_score=float(parsed.overall_score),
                feedback=str(parsed.feedback),
                details=details,
            )
        )
